scripts.py

```diff
@@ -707,6 +711,134 @@ def do_run(self):
+        if existing_timestamp:
+            # Get migration directory path for core and its containing server
+            current_dir = os.path.split(os.path.abspath(__file__))[0]
+            core = os.path.join(current_dir, 'migration')
+            server = os.path.join(os.path.split(current_dir)[0], 'migration')
+            directories_by_priority = [core, server]
+
+            # Get all migrations from both directories.
+            for directory in directories_by_priority:
+                dir_migrations = self.migration_files(os.listdir(directory))
+                migrations += dir_migrations
+                migrations_by_dir[directory] = dir_migrations
+
+            # Find any new migrations and run them.
+            new_migrations = self.get_new_migrations(
+                existing_timestamp, migrations
+            )
+            if new_migrations:
+                print "%d new migrations found." % len(new_migrations)
+                for migration in new_migrations:
+                    print "- %s" % migration
+
+                self.run_migrations(
+                    new_migrations, directories_by_priority, migrations_by_dir,
+                    existing_timestamp
+                )
+                print "All new migrations have been run."
+
+                # Update timestamp for the last run migration.
+                self.update_timestamp(existing_timestamp, new_migrations[-1])
+            else:
+                print "No new migrations found. Your database is up-to-date."
+        else:
+            print ''
+            print (
+                "NO TIMESTAMP FOUND. Run script with timestamp that indicates"
+                " the last migration run against this database."
+            )
+            self.argparser.print_help()
+
+    def migration_files(self, filelist):
+        """Filter a list of files for migration file extensions"""
+
+        migratable = [f for f in filelist
+            if (f.endswith('.py') or f.endswith('.sql'))]
+        return sorted(migratable)
+
+    def get_new_migrations(self, timestamp, migrations):
+        """Return a list of migration filenames, created since the timestamp
+        """
+        last_run = timestamp.timestamp.strftime('%Y%m%d')
+        new = [migration for migration in migrations
+                if int(migration[0:8]) >= int(last_run)]
+        return sorted(new)
+
+    def run_migrations(self, migrations, directories, migrations_by_dir,
+                       timestamp):
+        """Run the migrations in date order"""
+        previous = ''
+
+        for migration_file in sorted(migrations):
+            for d in directories:
+                if migration_file in migrations_by_dir[d]:
+                    # Create timestamp for previous migration, in case of
+                    # error while running this migration.
+                    if previous:
+                        set_trace()
+                        self.update_timestamp(timestamp, previous)
```

Comment:
The timestamp being set as the first time the date was run wouldn't be a huge problem right now because migrations are run if they are later than _or_ the same date as the provided timestamp. So if there was a failure between `20160810-2` and `20160810-3`, the whole date block (-1 to -n) would be run again.

It's not ideal, e.g. if there's a situation when someone creates a python migration that should only be run once for some reason (???), but I don't know where I could put the number otherwise without changing the Timestamp schema with an `extra` column or some such. Or maybe as hours in the datetime, lolomgno?